oai-pmh-handler/src/main/java/no/sikt/nva/oai/pmh/handler/GraphRecordTransformer.java

```diff
@@ -0,0 +1,182 @@
+package no.sikt.nva.oai.pmh.handler;
+
+import static java.util.Objects.nonNull;
+import static org.apache.jena.riot.Lang.JSONLD;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.io.ByteArrayInputStream;
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import no.unit.nva.commons.json.JsonUtils;
+import nva.commons.core.ioutils.IoUtils;
+import org.apache.jena.query.QueryExecutionFactory;
+import org.apache.jena.query.QuerySolution;
+import org.apache.jena.rdf.model.ModelFactory;
+import org.apache.jena.riot.RDFDataMgr;
+import org.openarchives.oai.pmh.v2.HeaderType;
+import org.openarchives.oai.pmh.v2.MetadataType;
+import org.openarchives.oai.pmh.v2.OaiDcType;
+import org.openarchives.oai.pmh.v2.ObjectFactory;
+import org.openarchives.oai.pmh.v2.RecordType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class GraphRecordTransformer implements RecordTransformer {
+
+  private static final String SPARQL_QUERY =
+      IoUtils.stringFromResources(Path.of("publication.sparql"));
+  private static final Logger LOGGER = LoggerFactory.getLogger(GraphRecordTransformer.class);
+  private static final ObjectFactory OBJECT_FACTORY = new ObjectFactory();
+  private static final String JSON_LD_GRAPH = "@graph";
+  private static final String JSON_LD_CONTEXT = "@context";
+  private static final String TITLE = "title";
+  private static final String CONTRIBUTOR = "contributor";
+  private static final String ID = "id";
+  private static final String MODIFIED_DATE = "modifiedDate";
+  private static final String PUBLISHER = "publisher";
+  private static final String DATE = "date";
+  private static final String TYPE = "type";
+
+  public GraphRecordTransformer() {}
+
+  @Override
+  public List<RecordType> transform(List<JsonNode> hits) {
+    LOGGER.info(String.format("Transforming %d hits!", Objects.isNull(hits) ? 0 : hits.size()));
+    if (Objects.isNull(hits) || hits.isEmpty()) {
+      return Collections.emptyList();
+    }
+    var optimized = optimizeForGraph(hits);
+    return getResult(optimized);
+  }
+
+  private static JsonNode optimizeForGraph(List<JsonNode> hits) {
+    var contextNode = collectContext(hits);
+    var arrayNode = new ArrayNode(JsonNodeFactory.instance);
+    hits.forEach(arrayNode::add);
+    var jsonGraph =
+        new ObjectNode(
+            JsonNodeFactory.instance,
+            Map.of(JSON_LD_GRAPH, arrayNode, JSON_LD_CONTEXT, contextNode));
```

Comment:
We can probably just apply a context that states that "hits" is an alias for "@graph"; I don't know that this would be more efficient tbh.

Reply:
Probably we could, but I keep it as is for now.